### ISSUE TITLE
Sensitvities dud[n,p,w] not available in Python API

### ISSUE THREAD

**Issue by bknueven** [reaktoro/reaktoro#379]
I am reaching out because we would like the final Jacobian matrix for the equilibrium calculation for use in our open-source tool [WaterTAP](https://github.com/watertap-org/watertap).

Specifically, `dudn` is part of the `EquilibriumProps` class which is not exposed to Python. I am happy to write the Python bindings, but maybe this choice is intentional. Relatedly, would it be better for these sensitivities to be part of the `EquilibriumSensitivity`class?

**Comment by allanleal** [comment]
The `EquilibriumProps` class is a utility class used in `EquilibriumSolver`. Using it outside of this context may be cumbersome or may not work as expected.

`EquilibriumSensitivity` should contain what you need I think. The Jacobian of the equilibrium calculation can be viewed as the derivative of the output with respect to the input in the calculation, where output are the amounts of species `n = [n1, ..., nN]` (and also Lagrange multipliers, but I'll ignore these here) and the inputs are  `w = [T, P]`  and `c = [H, O, Na, Cl, Ca, ...]`, the amounts of elements. Note: there could be more input conditions in `w` depending on how you customized the `EquilibriumSolver` with `EquilibriumSpecs`.

* `EquilibriumSensitivity::dndw` returns the Jacobian of the species amounts at equilibrium with respect to the input conditions `w`.
* `EquilibriumSensitivity::dndc` returns the Jacobian of the species amounts at equilibrium with respect to the input element amounts `c`.

Since `n` is not an input to the equilibrium calculation, `EquilibriumSensitivity` has no derivatives with respect to `n`. 

How do you want to use Jacobian matrices in WaterTAP in the context of equilibrium calculations?

**Comment by bknueven** [comment]
Hi @allanleal, thanks for your detailed reply. I realized that I was confused about `n`, in that it is species amounts in the _outputs_. To achieve our desired result, we have added relevant "open conditions" to the `EquilibriumSpecs` and then an artificial constraint constraining it to its current value. The sensitivity we are looking for can then be retrieved with `EquilibriumSensitivity::dudw`.

> How do you want to use Jacobian matrices in WaterTAP in the context of equilibrium calculations?

WaterTAP solves water treatment trains utilizing Pyomo and Ipopt. Water chemistry often a limiting factor, and to date we have developed surrogate models for detailed water chemistry, see, e.g., [Amusat et al.](https://pubs.acs.org/doi/full/10.1021/acsestengg.3c00537). We have developed some of our own equilibrium unit models, but as pointed out in the implementation paper for Reaktoro's solver, you run into issues with traditional interior point methods.

However, after reading Reaktoro's documentation we believe we can do something much more elegant than surrogate models and use Reaktoro's equilibrium solver as an implicit function in our treatment train simulation/optimization problem, see [Parker et al.](https://doi.org/10.1016/j.compchemeng.2022.108042) for a description of the general approach.
